src/ZenohUTransport.cpp

```diff
@@ -9,4 +9,384 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
+#include <up-cpp/datamodel/serializer/UUri.h>
+#include <up-cpp/datamodel/serializer/Uuid.h>
+#include <up-cpp/datamodel/validator/UMessage.h>
+#include <up-cpp/datamodel/validator/UUri.h>
+
+#include <stdexcept>
+
 #include "up-transport-zenoh-cpp/ZenohUTransport.h"
+
+namespace uprotocol::transport {
+
+const char UATTRIBUTE_VERSION = 1;
+
+const uint32_t WILDCARD_ENTITY_ID = 0x0000FFFF;
+const uint32_t WILDCARD_ENTITY_VERSION = 0x000000FF;
+const uint32_t WILDCARD_RESOURCE_ID = 0x0000FFFF;
+
+using namespace zenoh;
+using namespace uprotocol::v1;
+using namespace uprotocol::datamodel;
+
+UStatus ZenohUTransport::uError(UCode code, std::string_view message) {
+	UStatus status;
+	status.set_code(code);
+	status.set_message(std::string(message));
+	return status;
+}
+
+std::string ZenohUTransport::toZenohKeyString(
+    const std::string& default_authority_name, const UUri& source,
+    const std::optional<v1::UUri>& sink) {
+	std::ostringstream zenoh_key;
+
+	auto writeUUri = [&](const v1::UUri& uuri) {
+		zenoh_key << "/";
+
+		// authority_name
+		if (uuri.authority_name().empty()) {
+			zenoh_key << default_authority_name;
+		} else {
+			zenoh_key << uuri.authority_name();
+		}
+		zenoh_key << "/";
+
+		// ue_id
+		if (uuri.ue_id() == WILDCARD_ENTITY_ID) {
+			zenoh_key << "*";
+		} else {
+			zenoh_key << uuri.ue_id();
+		}
+		zenoh_key << "/";
+
+		// ue_version_major
+		if (uuri.ue_version_major() == WILDCARD_ENTITY_VERSION) {
+			zenoh_key << "*";
+		} else {
+			zenoh_key << uuri.ue_version_major();
+		}
+		zenoh_key << "/";
+
+		// resource_id
+		if (uuri.resource_id() == WILDCARD_RESOURCE_ID) {
+			zenoh_key << "*";
+		} else {
+			zenoh_key << uuri.resource_id();
+		}
+	};
+
+	zenoh_key << "up";
+	zenoh_key << std::uppercase << std::hex;
+
+	writeUUri(source);
+
+	if (sink.has_value()) {
+		writeUUri(*sink);
+	} else {
+		zenoh_key << "/{}/{}/{}/{}";
+	}
+	return zenoh_key.str();
+}
+
+std::vector<std::pair<std::string, std::string>>
+ZenohUTransport::uattributesToAttachment(const UAttributes& attributes) {
+	std::vector<std::pair<std::string, std::string>> res;
+
+	std::string version(&UATTRIBUTE_VERSION, 1);
+
+	std::string data;
+	attributes.SerializeToString(&data);
+
+	res.push_back(std::make_pair("", version));
+	res.push_back(std::make_pair("", data));
+	return res;
+}
+
+UAttributes ZenohUTransport::attachmentToUAttributes(
+    const AttachmentView& attachment) {
+	std::vector<BytesView> attachment_vec;
+	attachment.iterate(
+	    [&](const BytesView& key, const BytesView& value) -> bool {
+		    attachment_vec.push_back(value);
+		    return true;
+	    });
+
+	if (attachment_vec.size() != 2) {
+		// TODO: error report, exception?
+	}
+
+	if (attachment_vec[0].get_len() == 1) {
+		if (attachment_vec[0].as_string_view()[0] != UATTRIBUTE_VERSION) {
+			// TODO: error report, exception?
+		}
+	};
```

Comment:
We should evaluate what we want to do for errors here, but we probably don't want to throw. That should be reserved for unrecoverable errors. Receiving invalid data on the wire is not one of those situations. Maybe return `Expected<UAttributes, UStatus>`